library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/NamespaceResourceManagerAccessor.java

```
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.resource.loader.mixin;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Accessor;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+import net.minecraft.resource.NamespaceResourceManager;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.util.Identifier;
+
+@Mixin(NamespaceResourceManager.class)
+public interface NamespaceResourceManagerAccessor {
+	@Accessor("type")
+	ResourceType getType();
+
+	@Invoker("getMetadataPath")
+	static Identifier quilt$accessor_getMetadataPath(Identifier id) {
```

Mixin makes a guarantee that accessor Mixins will never have conflicting methods, the issue only arises when a duck exists that is named the same thing.

quilt prefix here is probably not needed.

---

IMO it's still a good idea to give these prefixes--the other major projects (Sponge, Fabric [iirc]) and their derivatives (Patchwork) use this style.

And in general it's best to avoid anything that could fail in Mixin because Mixin loves failing silently and it's awful to debug

---

Everytime I used an `Invoker` without a prefix I ended up with a crash because of "same name" somehow.

---

To clarify, the `quilt$` and `accessor_` are redundant here, one or the other has to go

---

I think I remember why the `accessor_` part is there, because I was the one who PRed this accessor to FAPI, and originally it was just `namespace$` prefixed, but I was told that I should've made it clear that it's an accessor and not a new method.

But the issue is, if I use `accessor_` and someone else put an invoker with the same `accessor_` prefix, won't I risk a crash?

---

Mixin seems to handle invokers properly with this aliasing thing, as many mods put accessors on very common things

---

Not anymore. That was fixed in mixin 0.8.3 iirc

---

accessors shouldn't ever conflict, when you got the crash were you using `accessor_getMetadataPath` or just `getMetadataPath`?

---

With just `getMetadataPath` it crashed I think?
It's been a while tbh so it might have changed since.

---

Yeah, you were conflicting with a non-invoker then